### PR TITLE
add zero optimizer parallel

### PR DESCRIPTION
**Pull request by CaitinZhao**
# What does this PR do?

参考deepspeed的zero，在MindSpore的数据并行模式下实现优化器并行算法

Fixes # (issue)

Adds # (feature)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/mindspore-lab/mindone/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the documentation with your changes? E.g. record bug fixes or new features in `What's New`. Here are the
      [documentation guidelines](https://github.com/mindspore-lab/mindcv/wiki/%E6%96%87%E6%A1%A3%E7%BC%96%E5%86%99%E8%A7%84%E8%8C%83)
- [ ] Did you build and run the code without any errors?
- [ ] Did you report the running environment (NPU type/MS version) and performance in the doc? (better record it for data loading, model inference, or training tasks)
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@xxx

### CODE REVIEW

File: tests/st/test_zero.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    if save_graph:
+        ms.set_context(save_graphs=True, save_graphs_path="ms_ir")
+    if distribute:
+        init()
+        group_size = get_group_size()
+        rank_id = get_rank()
+        print(f"rank_id: {rank_id}, group_size: {group_size}")
+        ms.reset_auto_parallel_context()
+        ms.set_auto_parallel_context(
+            parallel_mode=ms.ParallelMode.DATA_PARALLEL,
```
Comment (SamitHuang) [inline tests/st/test_zero.py:28]: How is the difference between it ParallelMode.SEMI_AUTO_PARALLEL?

Comment (CaitinZhao) [reply tests/st/test_zero.py:28]: Implemented a new version of zero optimizer parallel in  DATA_PARALLEL refer to DeepSpeed. Not use the MindSpore automatic parallel process.

**Review by SamitHuang — APPROVED** [review]
Suggestion: supplement the script for checkpoint merging.

File: mindone/trainers/train_step.py [hunk 2]
```diff
@@ -104,6 +115,10 @@ def construct(self, *inputs):
 
         # 1. compute gradients (of the up-scaled loss w.r.t. the model weights)
         grads = self.grad(self.network, weights)(*inputs, scaling_sens_filled)
+
+        # Gradient communication
+        grads = self.zero_helper.cal_gradients(grads)
```
Comment (zhtmike) [inline mindone/trainers/train_step.py:120]: `self.zero_helper` can be None, need to to add a if condition

Comment (CaitinZhao) [reply mindone/trainers/train_step.py:120]: ok

**Review by vigo999 — APPROVED** [review]

**Pull request merged**